### PR TITLE
chain: Reporter list of validator CLI and REST endpoint

### LINKED ISSUE #1886: Reporter list of validator REST endpoint
User can query a list of reporters of the given validator at `/oracle/reporters/<validator_address>` to return list of report address.

### PR DESCRIPTION
**Pull request by perimeko**
fixed #1886

### CODE REVIEW

File: chain/x/oracle/client/rest/rest.go [hunk 1]
```diff
@@ -21,4 +22,5 @@ func RegisterRoutes(cliCtx context.CLIContext, r *mux.Router, storeName string)
 	r.HandleFunc(fmt.Sprintf("/%s/oracle_scripts/{%s}", storeName, idTag), getOracleScriptByIDHandler(cliCtx, storeName)).Methods("GET")
 	r.HandleFunc(fmt.Sprintf("/%s/requests/{%s}", storeName, idTag), getRequestByIDHandler(cliCtx, storeName)).Methods("GET")
 	r.HandleFunc(fmt.Sprintf("/%s/request_search", storeName), getRequestSearchHandler(cliCtx, storeName)).Methods("GET")
+	r.HandleFunc(fmt.Sprintf("/%s/reporters/{%s}", storeName, validatorAddressTag), getAllReportersByValidatorAddressHandler(cliCtx, storeName)).Methods("GET")
```
Comment (sorawit) [inline chain/x/oracle/client/rest/rest.go]: Let's just call this `GetReportsHandler`. More concise.

File: chain/x/oracle/keeper/reporter.go [hunk 2]
```diff
@@ -33,3 +33,22 @@ func (k Keeper) RemoveReporter(ctx sdk.Context, val sdk.ValAddress, addr sdk.Acc
...
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		key := iterator.Key()
+		reporterBytes := key[21:]
+		reporterAddress := sdk.AccAddress(reporterBytes)
+		reporters = append(reporters, reporterAddress)
+	}
+
+	// Appends self reporter of validator to the list
+	selfReporter := sdk.AccAddress(val)
```
Comment (sorawit) [inline chain/x/oracle/keeper/reporter.go]: Can we add self first? I think having self as the first element makes more sense.

File: chain/x/oracle/keeper/reporter.go [hunk 3]
```diff
@@ -33,3 +33,22 @@ func (k Keeper) RemoveReporter(ctx sdk.Context, val sdk.ValAddress, addr sdk.Acc
...
+
+// GetAllReportersOfValidator returns the reporter list of the given validator.
+func (k Keeper) GetAllReportersOfValidator(ctx sdk.Context, val sdk.ValAddress) (reporters []sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	storeKeyPrefix := append(types.ReporterStoreKeyPrefix, val.Bytes()...)
+	iterator := sdk.KVStorePrefixIterator(store, storeKeyPrefix)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		key := iterator.Key()
+		reporterBytes := key[21:]
```
Comment (sorawit) [inline chain/x/oracle/keeper/reporter.go]: Can we use `1 + len(val)` or something like that? I prefer that more than hardcoding number 21.

**Review by sorawit — CHANGES_REQUESTED** [review]
We should also add CLI command together with this PR.

File: chain/x/oracle/client/rest/rest.go (hunk 1 above)
Comment (perimeko) [reply chain/x/oracle/client/rest/rest.go]: Could we change to `GetReportersHandler` ?

File: chain/x/oracle/keeper/reporter_test.go [hunk 4]
```diff
@@ -42,3 +43,24 @@ func TestRemoveReporter(t *testing.T) {
 	require.False(t, k.IsReporter(ctx, Alice.ValAddress, Bob.Address))
 	require.False(t, k.IsReporter(ctx, Alice.ValAddress, Carol.Address))
 }
+
+func TestGetAllReporter(t *testing.T) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/reporter_test.go]: GetAllReporters

File: chain/x/oracle/types/querier.go [hunk 5]
```diff
@@ -23,3 +26,8 @@ type QueryRequestResult struct {
 	Reports []Report `json:"reports"`
 	Result  *Result  `json:"result"`
 }
+
+// QueryReportersResult is the struct for the result of reporters query.
+type QueryReportersResult struct {
```
Comment (sorawit) [inline chain/x/oracle/types/querier.go]: I don't think this type is needed.

File: chain/x/oracle/client/cli/query.go [hunk 6]
```diff
@@ -141,3 +142,21 @@ func GetQueryCmdRequestSearch(route string, cdc *codec.Codec) *cobra.Command {
...
+	return &cobra.Command{
+		Use:  "reporters [validator]",
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cliCtx := context.NewCLIContext().WithCodec(cdc)
+			res, _, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/%s/%s", route, types.QueryReporters, args[0]), nil)
+			if err != nil {
+				return err
+			}
+			var out types.QueryReportersResult
```
Comment (sorawit) [inline chain/x/oracle/client/cli/query.go]: var out `[]sdk.AccAddress`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/client/cli/query.go [hunk 7]
```diff
@@ -141,3 +143,21 @@ func GetQueryCmdRequestSearch(route string, cdc *codec.Codec) *cobra.Command {
 		},
 	}
 }
+
+// GetQueryCmdReportersofValidator implements the query reporter list of validator command.
```
Comment (sorawit) [inline chain/x/oracle/client/cli/query.go]: Update comment.

File: chain/x/oracle/keeper/querier.go [hunk 8]
```diff
@@ -111,3 +113,16 @@ func queryRequestByID(ctx sdk.Context, path []string, k Keeper) ([]byte, error)
...
+func queryReporters(ctx sdk.Context, path []string, k Keeper) ([]byte, error) {
+	if len(path) != 1 {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "validator address not specified")
+	}
+
+	validatorAddress, err := sdk.ValAddressFromBech32(path[0])
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, err.Error())
+	}
+	reporters := k.GetReporters(ctx, validatorAddress)
```
Comment (sorawit) [inline chain/x/oracle/keeper/querier.go]: Let's pack this and the following line into just one line.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/client/cli/query.go [hunk 9]
```diff
@@ -12,6 +12,7 @@ import (
 
 	clientcmn "github.com/bandprotocol/bandchain/chain/x/oracle/client/common"
 	"github.com/bandprotocol/bandchain/chain/x/oracle/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
```
Comment (taobun) [inline chain/x/oracle/client/cli/query.go]: Bring it to line 12

**Review by sorawit — APPROVED** [review]

File: chain/x/oracle/keeper/reporter.go [hunk 10]
```diff
@@ -33,3 +33,22 @@ func (k Keeper) RemoveReporter(ctx sdk.Context, val sdk.ValAddress, addr sdk.Acc
...
 }
+
+// GetReporters returns the reporter list of the given validator.
+func (k Keeper) GetReporters(ctx sdk.Context, val sdk.ValAddress) (reporters []sdk.AccAddress) {
+	// Appends self reporter of validator to the list
+	selfReporter := sdk.AccAddress(val)
+	reporters = append(reporters, selfReporter)
+
+	store := ctx.KVStore(k.storeKey)
+	storeKeyPrefix := append(types.ReporterStoreKeyPrefix, val.Bytes()...)
```
Comment (taobun) [inline chain/x/oracle/keeper/reporter.go]: Move this line to `types/key.go` create a new function `ValidatorReporterPrefixKey` to generate this key

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**